blueocean-core-js/src/js/jwt.js

```diff
@@ -0,0 +1,96 @@
+import fetch from 'isomorphic-fetch';
+import jwt from 'jsonwebtoken';
+import UrlUtils from './urlconfig';
+import { FetchFunctions } from './fetch';
+import { jwk2pem } from 'pem-jwk';
+let storedToken = null;
+let publicKeyStore = null;
+let tokenFetchPromise = null;
+export default {
+    /**
+     * Fetches the JWT token. This token is cached for a default of 25mins.
+     * If it is within 5mins or expiry it will fetch a new one.
+     */
+    fetchJWT() {
+        if (storedToken && storedToken.exp) {
+            const diff = storedToken.exp - Math.trunc(new Date().getTime() / 1000);
+            if (diff < 300) {
+                tokenFetchPromise = null;
+            }
+        }
+
+        if (!tokenFetchPromise) {
+            tokenFetchPromise = fetch(`${UrlUtils.getJenkinsRootURL()}/jwt-auth/token`, { credentials: 'same-origin' })
+                .then(this.checkStatus)
+                .then(response => {
+                    const token = response.headers.get('X-BLUEOCEAN-JWT');
+                    if (token) {
+                        return token;
+                    }
+                    
+                    throw new Error('Could not fetch jwt_token');
```

Comment:
If token fetch gets a 401, they MUST login. This will be taken care of in a future PR.

If the jenkins instance allows anonymous access, they will still get a token.